File: services/go/graphql/src/domain/interfaces/product_repository.go
package interfaces

import "graphql/src/domain/entities"

type ProductRepository interface {
	GetByID(id int) (*entities.Product, error)
	GetAll() ([]entities.Product, error)
	Create(input entities.CreateInput) (*entities.Product, error)
	Update(id int, input entities.UpdateInput) (*entities.Product, error)
	Delete(id int) (bool, error)
	DeleteAll() (bool, error)
}

File: services/rust/grpc/src/domain/interfaces/product_repository.rs
use async_trait::async_trait;

use crate::domain::entities::product::{CreateProductInput, Product, UpdateProductInput};

#[async_trait]
pub trait ProductRepository: Send + Sync {
    async fn get_by_id(&self, id: i32) -> Result<Option<Product>, sqlx::Error>;
    async fn get_all(&self) -> Result<Vec<Product>, sqlx::Error>;
    async fn create(&self, request: CreateProductInput) -> Result<Product, sqlx::Error>;
    async fn update(&self, id: i32, request: UpdateProductInput) -> Result<Product, sqlx::Error>;
    async fn delete(&self, id: i32) -> Result<bool, sqlx::Error>;
    async fn delete_all(&self) -> Result<bool, sqlx::Error>;
}
